pkg/securelaunch/helpers.go

```diff
@@ -0,0 +1,270 @@
+
+package securelaunch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/cmdline"
+	"github.com/u-root/u-root/pkg/diskboot"
+	"github.com/u-root/u-root/pkg/find"
+	"github.com/u-root/u-root/pkg/mount"
+	"github.com/u-root/u-root/pkg/storage"
+)
+
+/* used to store all block devices returned from a call to storage.GetBlockStats */
+var storageBlkDevices []storage.BlockDev
+
+/*
+ * if kernel cmd line has uroot.uinitargs=-d, debug fn is enabled.
+ * kernel cmdline is checked in sluinit.
+ */
+var Debug = func(string, ...interface{}) {}
+
+/*
+ * modified ver of https://stackoverflow.com/questions/10510691/how-to-check-whether-a-file-or-directory-exists
+ * exists returns whether the given file or directory exists
+ * also returns if the path is a directory
+ */
+func exists(path string) (bool, bool, error) {
+	fileInfo, err := os.Stat(path)
+	if err == nil {
+		return true, fileInfo.IsDir(), nil
+	}
+	if os.IsNotExist(err) {
+		return false, false, nil
+	}
+	// file or dir def exists..
+	return true, fileInfo.IsDir(), err
+}
+
+/*
+ * data is a []byte slice
+ * dst is an already mounted file path.
+ * dst is an Absolute file path.
+ * defFileName is default dst file name, only used if user doesn't provide one.
+ * returns
+ * 1. target file path where file was written
+ * 2. error
+ */
+func WriteToFile(data []byte, dst, defFileName string) (string, error) {
+
+	// make sure dst is an absolute file path
+	if !filepath.IsAbs(dst) {
+		return "", fmt.Errorf("Error: dst =%s Not an absolute path ", dst)
+	}
+
+	// target is the full absolute path where []byte will be written to
+	target := dst
+
+	/* dst = /foo/bar/eventlog, check if /foo/bar/eventlog is a dir or file that exists..
+	     * if it exists as a file, this if loop is untouched.
+		 * if it exists as a dir, append defFileName to it
+	     * NOTE both dst = /foo/bar/ and dst=/foo/bar are considered dirs i.e. trailing "/" has no meaning..
+	*/
+	dstFound, is_dir, err := exists(dst)
+	if is_dir {
+		if !dstFound {
+			return "", fmt.Errorf("destination dir doesn't")
+		} else {
+			// User provided a dir, Use defFileName as target
+			Debug("No file name provided. Adding it now. old target=%s", target)
+			target = filepath.Join(dst, defFileName)
+			Debug("New target=%s", target)
+		}
+	}
+
+	Debug("target=%s", target)
+	err = ioutil.WriteFile(target, data, 0644)
+	if err != nil {
+		return "", fmt.Errorf("Could't write file to %s, err=%v", target, err)
+	}
+	Debug("WriteToFile exit w success data written to target=%s", target)
+	return target, nil
+}
+
+/*
+ * src and dst are already mounted file paths.
+ * src and dst are Absolute file paths
+ * defFileName is default dst file name, only used if user doesn't provide one.
+ * returns
+ * 1. target file path where file was written
+ * 2. error
+ */
+func WriteSrcFiletoDst(src, dst, defFileName string) (string, error) {
+	srcFileFound, _, err := exists(src)
```

Comment:
You can't contract all this, and get better error messages if you do the following:

```go
d, err := ioutil.ReadFile(src)
if err != nil {
  return fmt.Errorf("could not copy file: %v", err)
}
```

stdlib path-based file system functions will already return `os.PathError` which include the src path, no need to use fmt.Errorf to augment the path.

ReadFile will also say if the file doesn't exist, so no need to check again.

Reply:
see https://play.golang.org/p/Yc0PxnjRvmH